plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -525,15 +528,15 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
                                          bool return_failure_traces,
                                          next_function<transaction_trace_ptr> next) {
          chain::controller& chain = chain_plug->chain();
-         const auto max_trx_time_ms = _max_transaction_time_ms.load();
+         const auto max_trx_time_ms = ( trx_type == transaction_metadata::trx_type::read_only ) ? _max_read_only_transaction_time_ms : _max_transaction_time_ms.load();
```

Comment:
`_max_read_only_transaction_time_ms` is not thread safe. It would need to be made `atomic` to be used here as this is called from the net_plugin thread pool. However, you need to access it here as you can use `max_trx_time_ms` of `-1` for read only transactions since `start_recover_keys` doesn't do anything for read only because read only do not have signatures.

Reply:
`transaction::get_signature_keys` should not call `sig_digest` when `signatures.empty()`. This will be a slight optimization for read only transactions.

Reply:
Thanks for those two suggestions. Will change.